### PR TITLE
Implement tag dispatching for sort algorithms

### PR DESCRIPTION
**Pull request by dparanic**

### CODE REVIEW

File: pstl/include/pstl/internal/execution_impl.h [hunk 1]
```diff
@@ -120,11 +120,13 @@ struct __tbb_backend
...
 #ifdef _PSTL_PAR_BACKEND_TBB
     __tbb_backend;
-#elif _PSTL_PAR_BACKEND_SERIAL
+#else
+#ifdef _PSTL_PAR_BACKEND_SERIAL
     __serial_backend;
 #else
 #    error "A parallel backend must be specified";
 #endif
+#endif
```
Comment (rarutyun) [inline pstl/include/pstl/internal/execution_impl.h]:
```cpp
#if defined(_PSTL_PAR_BACKEND_TBB)
    __tbb_backend;
#elif defined(_PSTL_PAR_BACKEND_SERIAL)
    __serial_backend;
#else
#    error "A parallel backend must be specified";
#endif
```

File: pstl/include/pstl/internal/algorithm_impl.h [hunk 2]
```diff
@@ -2248,9 +2251,9 @@ __pattern_partial_sort_copy(_ExecutionPolicy&& __exec, _RandomAccessIterator1 __
             typedef typename std::iterator_traits<_RandomAccessIterator1>::value_type _T1;
             typedef typename std::iterator_traits<_RandomAccessIterator2>::value_type _T2;
             __par_backend::__buffer<_T1> __buf(__n1);
-            _T1* __r = __buf.get();
+            _T1* __r = __buf.get();
```
Comment (rarutyun) [inline pstl/include/pstl/internal/algorithm_impl.h]: seems like extra space at the end

**Review by rarutyun — COMMENTED** [review]
Other than small improvement suggestions looks good to me.

File: pstl/include/pstl/internal/execution_impl.h (hunk 1 above)
Comment (dparanic) [reply pstl/include/pstl/internal/execution_impl.h]: Applied.

File: pstl/include/pstl/internal/algorithm_impl.h (hunk 2 above)
Comment (dparanic) [reply pstl/include/pstl/internal/algorithm_impl.h]: Deleted.

File: pstl/include/pstl/internal/glue_algorithm_impl.h [hunk 3]
```diff
@@ -111,12 +111,10 @@ __pstl::__internal::__enable_if_execution_policy<_ExecutionPolicy, _ForwardItera
 find_end(_ExecutionPolicy&& __exec, _ForwardIterator1 __first, _ForwardIterator1 __last, _ForwardIterator2 __s_first,
          _ForwardIterator2 __s_last, _BinaryPredicate __pred)
 {
-    return __pstl::__internal::__pattern_find_end(
-        std::forward<_ExecutionPolicy>(__exec), __first, __last, __s_first, __s_last, __pred,
-        __pstl::__internal::__is_vectorization_preferred<_ExecutionPolicy, _ForwardIterator1, _ForwardIterator2>(
-            __exec),
-        __pstl::__internal::__is_parallelization_preferred<_ExecutionPolicy, _ForwardIterator1, _ForwardIterator2>(
-            __exec));
+    auto __dispatch_tag = __pstl::__internal::__select_backend(__exec, __first);
```
Comment (rarutyun) [inline pstl/include/pstl/internal/glue_algorithm_impl.h]: you should pass each iterator representative for individual type. `__first, __s_first)` would be correct. Please check for all the places of `__select_backend` usage.

File: pstl/test/support/utils.h [hunk 4]
```diff
@@ -231,7 +231,8 @@ fill_data(Iterator first, Iterator last, F f)
     }
 }
 
-struct MemoryChecker {
+struct MemoryChecker
```
Comment (rarutyun) [inline pstl/test/support/utils.h]: There are changes that are unrelated to functionality.

File: pstl/include/pstl/internal/algorithm_fwd.h [hunk 5]
```diff
@@ -983,18 +958,17 @@ _OutputIterator __brick_merge(_RandomAccessIterator1, _RandomAccessIterator1, _R
                               _RandomAccessIterator2, _OutputIterator, _Compare,
                               /* __is_vector = */ std::true_type) noexcept;
 
-template <class _ExecutionPolicy, class _ForwardIterator1, class _ForwardIterator2, class _OutputIterator,
-          class _Compare, class _IsVector>
+template <class _Tag, class _ExecutionPolicy, class _ForwardIterator1, class _ForwardIterator2, class _OutputIterator,
```
Comment (rarutyun) [inline pstl/include/pstl/internal/algorithm_fwd.h]: pattern_merge should be separated. Don't do the review too big, we lose granularity.

File: pstl/include/pstl/internal/glue_algorithm_impl.h (hunk 3 above)
Comment (dparanic) [reply pstl/include/pstl/internal/glue_algorithm_impl.h]: Done.

File: pstl/include/pstl/internal/glue_algorithm_impl.h [hunk 6]
```diff
@@ -799,12 +787,10 @@ __pstl::__internal::__enable_if_execution_policy<_ExecutionPolicy, _RandomAccess
 partial_sort_copy(_ExecutionPolicy&& __exec, _ForwardIterator __first, _ForwardIterator __last,
                   _RandomAccessIterator __d_first, _RandomAccessIterator __d_last, _Compare __comp)
 {
-    return __pstl::__internal::__pattern_partial_sort_copy(
-        std::forward<_ExecutionPolicy>(__exec), __first, __last, __d_first, __d_last, __comp,
-        __pstl::__internal::__is_vectorization_preferred<_ExecutionPolicy, _ForwardIterator, _RandomAccessIterator>(
-            __exec),
-        __pstl::__internal::__is_parallelization_preferred<_ExecutionPolicy, _ForwardIterator, _RandomAccessIterator>(
-            __exec));
+    auto __dispatch_tag = __pstl::__internal::__select_backend(__exec, __first);
```
Comment (rarutyun) [inline pstl/include/pstl/internal/glue_algorithm_impl.h]: __d_first is missed

File: pstl/include/pstl/internal/algorithm_fwd.h (hunk 5 above)
Comment (dparanic) [reply pstl/include/pstl/internal/algorithm_fwd.h]: Detached.

File: pstl/test/support/utils.h (hunk 4 above)
Comment (dparanic) [reply pstl/test/support/utils.h]: Removed.

File: pstl/include/pstl/internal/glue_algorithm_impl.h (hunk 6 above)
Comment (dparanic) [reply pstl/include/pstl/internal/glue_algorithm_impl.h]: Fixed.

**Review by rarutyun — APPROVED** [review]
